pyop2/types/dat.py

```diff
@@ -492,6 +492,20 @@ def norm(self):
         from math import sqrt
         return sqrt(self.inner(self).real)
 
+    def axpy(self, alpha: float, other: 'Dat') -> None:
```

Comment:
Much better having it hidden away in pyop2!

Please could you also add the `maxpy` (_multiple_ axpy) routine that we discussed in the meeting and use this in `assemble`?
`maxpy` takes in a sequence of alphas and a sequence of xs and increments them all into y.
For now it could just be implemented as a loop over axpy for each (alpha,x) pair, but in the future it could allow better control over data movement to/from GPUs.

Reply:
I like the idea of having a `maxpy` like this:
```
def maxpy(self, scalar: tuple, other: tuple) -> None:
        """Compute a sequence of axpy operations.

        This is equivalent to calling :meth:`axpy` for each pair of
        scalars and :class:`Dat` in the input sequences.

        :arg scalar: a sequence of scalars
        :arg other: a sequence of :class:`Dat`\s

        See also :meth:`axpy`.
    
        """
        for alpha, dat in zip(scalar, other):
            self.axpy(alpha, dat)
```
Any strong opinion against it?

Reply:
Personally I think I prefer to, instead of passing
```
dat.maxpy([2.9, 3.3], [datA, datB])
```
to have
```
dat.maxpy([(2.9, datA), (3.3, datB)])
```

Reply:
@Ig-dolci I think for now, that implementation is fine, it's what we are doing anyway, and encapsulating it means we can optimise later if it shows up on profiles.

@connorjward I'm not sure I have a preference either way. The current signature is closer to PETSc Vecs https://petsc.org/release/manualpages/Vec/VecMAXPY/ but your suggestion is probably more "pythonic". Although you wouldn't need to pass them as a list, it could just be `dat.maxpy((a0, x0), (a1, x1), ...)` and splat it into `*args`.

Reply:
Indeed your suggestion is the most "pythonic" of all!

In general I think it is not a great idea to try and mimic a C API in Python. This is something that really winds me up about petsc4py.

Reply:
Ha, fair enough. Pairs of (alpha, x) probably best.

Reply:
I think most use cases would require to unpack and repack the two lists of scalars and dats into the list of tuples. Are we sure about this?

Reply:
I guess the difference in the actual example we use above would be:
```py
result.dat.maxpy(expr.weights(), [a.dat for a in args])
```
vs
```py
result.dat.maxpy(*(zip(expr.weights(), [a.dat for a in args])))
```
so maybe you're right

Reply:
Thinking in the most used cases of two lists of scalars and dats. I would prefer `maxpy(scalar, x)`.